Develop

wch-mirek opened the pull request:
Made changes accordingly to feedback from [Redbooth task](https://redbooth.com/a/#!/projects/2135400/tasks/48960839)

wch-mirek committed:
eslint from AirBnb added and adjusted code to those rules
Modified "archiveAndPaw" to "buildAndPaw" => new naming is more clear

wch-mirek committed:
Examples from official OpenAPI repo
More body types handleld
More auth types handled

wch-mirek committed:
Missing Paw types definitions

wch-mirek committed:
Set "Bearer auth" handler as empty function as this do not need any extra work to be done to parse it

wch-mirek committed:
Adjusted body converter

wch-mirek committed:
Adjusted body converter

wch-mirek committed:
Issue fix

wch-mirek committed:
Replace URL parameters with default values -> problem from example "uspto"

mittsh commented on src/index.ts:
```diff
@@ -1,62 +1,67 @@
-import Paw from "./types-paw-api/paw"
-import OpenAPI, {NonRequiredLabel, BasicCredentialsLabel} from "./types-paw-api/openapi"
-import OpenAPIToPawConverter from "./lib/openapi-to-paw-converter/openapi-to-paw-converter";
+/* eslint-disable */
```
Why disabling eslint altogether in this file?

wch-mirek replied:
This line was added by my PHP Storm while executing "Fix whole file" options from ESLint and I haven't noticed that this line was added.
My mistake - sorry for that - I removed that in recent commit

mittsh commented on src/lib/openapi-to-paw-converter/components/auth-converter.ts:
```diff
@@ -1,94 +1,134 @@
+
+    if (authIn === 'query') {
+      value = this.request.getUrlParameterByName(authName) as string;
+      this.request.addUrlParameter((authName || ''), (value || ''));
```
I don't understand the logic here: query a URL param from `this.request` and them import it again to the same `this.request`? Shouldn't it scan `this.openApi.components.examples`?

wch-mirek replied:
That is correct - that logic make no sense - I noticed that I tried this in Generator as well and I noticed that in other repo, but didn't fixed that here also.

In recent commit I added a check and a comment that inform developer that authorizations made by http bearer or apiKey should be already parsed by ParametersConverter (as those authorization are made by sending header, cookie or query param).

About scanning examples - I think that shouldn't be done as You made a good point in [this](https://github.com/luckymarmot/Paw-OpenAPI3Generator/pull/1#discussion_r479772651) comment (no authorization details should be available in docs files).

mittsh replied:
Got it, thanks for clarifying @wch-mirek

mittsh replied:
Got it, thanks for clarifying @wch-mirek

mittsh commented on src/lib/openapi-to-paw-converter/components/auth-converter.ts:
```diff
@@ -1,94 +1,134 @@
+                this.parseOAuth2Auth(securityScheme.flows);
               } else {
-                throw new Error('No security match found')
+                throw new Error('No security match found');
```
Are we sure to cover all security schemes supported by the spec (`basic`, `bearer`, `apiKey`, `oauth2`)? We don't want to throw an error and interrupt the import if we miss a scheme.

My suggestion:
1. If you're 100% sure that these are the only possible values, then it's ok to keep the `throw`
2. If not, replace this by `console.error()` or `console.warn()` so we can always see that in the console and continue the import process

wch-mirek replied:
Without solid testing I'll never be 100% sure about this - in recent commit I changed that `throw` to `console.error` :)

mittsh commented on src/lib/openapi-to-paw-converter/components/auth-converter.ts:
```diff
@@ -1,94 +1,134 @@
+        const flow = flows[grantType] as OpenAPI.OAuthFlowObject;
+
+        this.request.oauth2 = {
+          client_id: '',
```
Forgive my ignorance here: there is no way to obtain "default" values for Client ID, Secret, and Token?

wch-mirek replied:
The oauth2 object was built accordingly to one I found in Postman plugin: https://github.com/luckymarmot/Paw-PostmanCollectionV2Importer/blob/a3e6a895a9642579f793c09eb4e361bcc7a5e427/src/lib/convertAuth.ts#L88 

Empty/undefined values were assigned because of TypeScript definition for OAuth2 object

mittsh replied:
I think that's actually more OpenAPI specific — whether the default values for OAuth 2 are provided in an OpenAPI 3.0 file. And I think they aren't. Correct me if I'm wrong 😄

wch-mirek replied:
I'm getting a little confused in that place :)

Your first comment was a question about "can we get some default values for OAuth2?" and Your second comment was "OpenAPI do not provide any OAuth2 default values" - am I correct here or maybe I misunderstood some details from Your comments?

As a conclusion: we cannot fetch any OAuth2 authorization values from OpenAPI file - [here](https://swagger.io/specification/#oauth-flows-object) is docs to corresponding FlowObject from OpenAPI

mittsh commented on src/lib/openapi-to-paw-converter/components/body-converter.ts:
```diff
@@ -0,0 +1,65 @@
+                break;
+              case 'application/x-www-form-urlencoded':
+                this.request.urlEncodedBody = (
+                  BodyConverter.parseToJsonIfString(body) as MapKeyedWithString<string>
```
Are we sure that the body is JSON-encoded when in URL Encode (`application/x-www-form-urlencoded`) and Multipart (`multipart/form-data`) modes?

wch-mirek replied:
The body variable can contain "any", so that parseToJsonIfString() function should have more complex logic than just parse stringified JSON to JSON object. To cover that topic it would be best to have as many possible examples as we can get.

About Paw's request urlEncodedBody and multipartBody params - they should contain JSON ([multipart](https://github.com/luckymarmot/Paw-PostmanCollectionV2Importer/blob/a3e6a895a9642579f793c09eb4e361bcc7a5e427/src/types-paw-api/paw.d.ts#L174), [URL encoded](https://github.com/luckymarmot/Paw-PostmanCollectionV2Importer/blob/a3e6a895a9642579f793c09eb4e361bcc7a5e427/src/types-paw-api/paw.d.ts#L167)). 
Postman Importer also converts Postman Collection to JSON while importing [URL encoded body](https://github.com/luckymarmot/Paw-PostmanCollectionV2Importer/blob/a3e6a895a9642579f793c09eb4e361bcc7a5e427/src/lib/convertBody.ts#L36) or [multipart body](https://github.com/luckymarmot/Paw-PostmanCollectionV2Importer/blob/a3e6a895a9642579f793c09eb4e361bcc7a5e427/src/lib/convertBody.ts#L49).

mittsh replied:
I'm not sure if I follow you here… 

> Postman Importer also converts Postman Collection to JSON while importing URL encoded body or multipart body.

It doesn't actually. If you look at [convertBodyUrlEncoded](https://github.com/luckymarmot/Paw-PostmanCollectionV2Importer/blob/a3e6a895a9642579f793c09eb4e361bcc7a5e427/src/lib/convertBody.ts#L36) or [convertBodyMultipart](https://github.com/luckymarmot/Paw-PostmanCollectionV2Importer/blob/a3e6a895a9642579f793c09eb4e361bcc7a5e427/src/lib/convertBody.ts#L49) it does not parse JSON in any way.

It only assigns a key-value/dict to the `urlEncodedBody` or `multipartBody` prop.

mittsh replied:
To clarify:

![image](https://user-images.githubusercontent.com/1319906/92236131-78bb6c80-eebd-11ea-8d2b-d632a65f853d.png)

wch-mirek replied:
That "No nested children" part lighten up what You mentioned here - I'll cover that case tomorrow as I focused on other topics with that extenstion.

wch-mirek replied:
I added a code to parse that body for urlencoded and multipart as key-value dictionary

mittsh commented on src/lib/openapi-to-paw-converter/components/body-converter.ts:
```diff
@@ -0,0 +1,65 @@
+  }
+
+  static parseToJsonIfString(value: any): object | MapKeyedWithString<string> {
+    return typeof value !== 'string' ? value : JSON.parse(value);
```
Can you wrap it in a `try/catch` in case there's an error and this is not a JSON? In which case, we can just leave the string.

Something like:

```ts
static parseToJsonIfString(value: any): object | MapKeyedWithString<string> {
  if (typeof value !== 'string') {
    return value
  }
  try {
    return JSON.parse(value)
  } catch (e) {
    console.error("Error parsing JSON body", e.message)
    return value;
  }
}
```

wch-mirek replied:
I wrapped it accordingly in recent commit

mittsh commented on src/lib/url.ts:
```diff
@@ -1,66 +1,101 @@
+      Object.entries(server.variables).forEach(([variableName, variable]) => {
+        const variableValue = variable.default ?? variableName;
+        newUrl = newUrl.replace(`{${variableName}}`, variableValue);
+      });
```
I think we can do something better here, with parsing the all the variables and creating a Request Variable in Paw for each one, unless you already have one before.

Something like this would be great: https://github.com/luckymarmot/Paw-PostmanCollectionV2Importer/blob/master/src/lib/convertEnvString.ts

Quickly adapting to request variables would look like:

```ts
const makeDs = (...components: Paw.DynamicStringComponent[]): DynamicString => {
  return new DynamicString(...components)
}

const getOrCreateRequestVariable = (varName: string, value: string, description: string, request: Paw.Request): DynamicValue => {
  let variable = request.getVariableByName(varName)
  if (!variable) {
    variable = request.addVariable(varName, value, description)
  }
  return variable.createDynamicValue()
}

const convertEnvString = (s: string, request: Paw.Request, context: Paw.Context): string|DynamicString => {
  const re = /\{\{([^}]+)\}\}/g
  let match
  const components: Paw.DynamicStringComponent[] = []
  let idx = 0

  // eslint-disable-next-line no-cond-assign
  while (match = re.exec(s)) {
    // push any string here before
    if (match.index > idx) {
      components.push(s.substr(idx, match.index - idx))
    }

    // push env variable
    // @TODO maybe here see if we can insert a logic to find a possible value for this variable
    components.push(getOrCreateRequestVariable(match[1] /* name */, '' /* value */, '' /* description */))

    idx = match.index + match[0].length
  }

  // add remaining string
  if (idx < s.length) {
    components.push(s.substr(idx))
  }

  // return
  if (components.length === 0) {
    return ''
  }
  if (components.length === 1 && typeof components[0] === 'string') {
    return components[0]
  }
  return makeDs(...components)
}
```

Forgive me if this code isn't perfect…I couldn't test it myself now. But feel free to ping me on Whatsapp or somewhere if I can help.

mittsh replied:
Actually reading over this again, `newUrl = newUrl.replace(`{${variableName}}`, variableValue);` is only working with server variables and not with request variables as can be found in the path parameters (`OpenAPI.ParameterObject` with `in === path`).

mittsh replied:
My recommendation: import the path parameters at the same time as you import the URL, as both are closely related 😊

wch-mirek replied:
I made some digging in code about that case and came up with some reflections.

The code part You pointed out was from URL class which actually should not do that - that code snippet You commented I added because of some issue which I'll describe a little later. Firstly I would suggest to not apply Your url parsing type, but stick to the one that [already exist](https://github.com/luckymarmot/Paw-OpenAPI3Importer/blob/61984a4e444cac4a01115a34bc4b61615c999e38/src/lib/openapi-to-paw-converter/components/parameters-converter.ts#L13) :) In my opinion that code from attachParametersFromOperationToRequest function do all parameters parsing in unified way for all types: query params, path params, headers and cookies. And that should work for all, but I came up with some problems and that is why I added additional param-replace code to URL class that You commented here.

The problem I came across is about putting variables into URL while importing to Paw.

_Case:_
The most edge case is the one from [uspto.json example](https://github.com/luckymarmot/Paw-OpenAPI3Importer/blob/61984a4e444cac4a01115a34bc4b61615c999e38/examples/openapi-official-examples/uspto.json#L77).
Full URL: {scheme}://developer.uspto.gov/ds-api/{dataset}/{version}/records

_Problems:_
All variables from URL must be replaced with request variables. The request variables from hostname in URL ({scheme}) should be replaced using OpenAPI's server variables and request variables from pathname in URL ({dataset}, {version}) should be replaced using OpenAPI's request variables. 
So to visualize - this is what I get for now:
<img width="682" alt="image" src="https://user-images.githubusercontent.com/69789351/92211197-58c18400-ee90-11ea-8b5f-66020b04004f.png">
and this is what should be:
<img width="695" alt="image" src="https://user-images.githubusercontent.com/69789351/92211268-7e4e8d80-ee90-11ea-96e3-6e412a28f777.png">

_Solution:_
For now I have no idea how to solve this - @mittsh do You know how I can import request variables into URL in Paw?

_Side note:_
For now I'm wondering why I actually added that replacing URL parameters with default values from server variables code, but I think it is because I tried to import upsto.json example and immediately export it using Generator -> and it is not working if {scheme} variable is not replaced :) This is the partial output with wrong data:
```
{
  "openapi": "3.0.3",
  "info": {
    "title": "Untitled Paw Document",
    "version": "1599204435388"
  },
  "paths": {
    "//developer.uspto.gov/ds-api/{dataset}/{version}/records/{scheme}/": {
      "servers": [
        {
          "url": "http://%7Bscheme%7D/"
        }
      ],
(...)
```

mittsh replied:
@wch-mirek 

> For now I have no idea how to solve this - @mittsh do You know how I can import request variables into URL in Paw?

Yes, that's exactly the problem I'm solving in my [code sample above](https://github.com/luckymarmot/Paw-OpenAPI3Importer/pull/1#discussion_r478450156):
1. It parses the URL from OpenAPI (e.g. `{scheme}://developer.uspto.gov/ds-api/{dataset}/{version}/records`)
2. Generates a `DynamicString` for Paw that's ready to be added to `request.url`

wch-mirek replied:
I take Postman Importer as an example and I used Environment Variables instead of Request Variables.

That URL parsing is happening before creating actual request in Paw, so I could not made URL with dynamic strings attached to Request Variables. And I haven't found a way to do so after request is created.

Using Environment Variables got me to same results as with Request Variables, so I think that solution is also acceptable?

wch-mirek replied:
I take Postman Importer as an example and I used Environment Variables instead of Request Variables.

That URL parsing is happening before creating actual request in Paw, so I could not made URL with dynamic strings attached to Request Variables. And I haven't found a way to do so after request is created.

Using Environment Variables got me to same results as with Request Variables, so I think that solution is also aceptable

mittsh replied:
No, this is not an acceptable solution. Env Variables are very different from Request Variables. The Postman Importer uses Env Variables because these are Env Variables in Postman.

As you may see, Request Variables in Paw contain a description, a schema, a required `true`/`false` property, which isn't possible as an Env Variable.

Additionally, Env Variables are, by design, global to the file. Whereas Request Variables are local to the request.

I would be happy to provide you with a solution to make this happen.

mittsh reviewed (commented):
Guys, this is really really cool 👏 Awesome progress since the last version, I'm happy to see that most body, params, auth formats are now supported.

I have left a few comments below, please see.

mittsh committed:
Use OpenAPI info.title

mittsh committed:
Add use of request vars in parseQueryParam()

mittsh committed:
Add JSON Schema in params

mittsh committed:
Add request variable description in parsePathParam()

mittsh commented on src/lib/openapi-to-paw-converter/components/parameters-converter.ts:
```diff
@@ -1,84 +1,90 @@
+    this.request.addHeader(
+      (param as OpenAPI.ParameterObject).name,
+      ParametersConverter.getValueFromParam(param),
+    );
```
Please import the description by using the same strategy as I've used here https://github.com/luckymarmot/Paw-OpenAPI3Importer/pull/2#discussion_r479741033 

Descriptions are important as we're importing from OpenAPI — which is an API _**Description**_ format 😄

wch-mirek replied:
I added that logic with converting schema to header, but it doesn't seem to work in my example ("exported-with-generator.yaml"). To quickly overcome this I changed lines:
```
param.description ?? '',
```
to
``` 
param.description ?? (param.schema as OpenAPI.SchemaObject)?.description ?? '',
```

and descriptions started to import correctly from my example file :)

mittsh replied:
As long as it works, I'm happy

wch-mirek committed:
Merge pull request #2 from luckymarmot/mittsh/develop-fixes

Develop Improvements

wch-mirek committed:
Minor fixes from PR

wch-mirek committed:
Better description adding logic in parameters-converter.ts

wch-mirek committed:
Parsing dynamic parameters using EnvironmentManager

wch-mirek committed:
Added OpenAPI 3.0 Sample Project.paw

wch-mirek committed:
Added urlencoded and multipart body parsing to force key-value dictionary

wch-kuba committed:
Rough fixes of importing env and request params. Will refactor this soon.

mittsh committed:
Fix crash importing variables

tmclaugh commented:
Any more progress on this?

pongstr committed:
chore(improve-importer): init

Co-authored-by: Micha Mazaheri <micha@mazaheri.me>

- updated: utilize @apidevtools/swagger-parser to validate and parse OpenAPI documents
- updated: utilize [openapi-types](https://github.com/kogosoftwarellc/open-api/tree/master/packages/openapi-types#readme) for typings
- chore: upgraded webpack, babel setup and jest setup
- chore: updated extension config references
- chore: simplify code coherence, reduce code complexity

**TODO**
- [ ] `utils/paw-url.ts` improve code coherence, utilize `URL` api
- [ ] `utils/json-schema-parser.ts` - improve code, ensure compliance with [openapi/json-schema](https://tools.ietf.org/html/draft-wright-json-schema-00#section-4.2) specs
- [ ] `utils/console.ts` - ensure to cover cyclic objects+references

Co-authored-by: Micha Mazaheri <micha@mazaheri.me>

pongstr committed:
chore(update-readme): updated README.md and LICENSE

mittsh committed:
chore(improvements): import server object

* Suggest a shorter name. Remove redundant “Importer” name.
* Fix import from URL
* Import server object variables and use them as env variables
